src/main/java/Duke.java

```diff
@@ -1,10 +1,107 @@
+    
+    /**
+     * The Storage object which loads and writes data to the hard drive.
+     */
+    private Storage storage;
+    
+    /**
+     * The Ui object which scans input and prints feedback to the user.
+     */
+    private Ui ui;
+    
+    /**
+     * Initializes the Ui, Storage, and TaskList objects.
+     *
+     * @param filePath The file path to write data to.
+     */
+    public Duke(String filePath) {
+        ui = new Ui();
+        storage = new Storage(filePath);
+        try {
+            taskList = new TaskList(storage.loadSavedList());
+        } catch (IOException e) {
+            ui.showError(e);
+            taskList = new TaskList();
+        }
+    }
+    
+    /**
+     * Scans and parses commands given by the user.
+     * Modifies the Tasks in the TaskList object based on the commands received by the user.
+     */
+    public void run() {
+        ui.printHello();
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                String input = ui.getNextLine();
+                String instruction = Parser.parseInstruction(input);
+                if (instruction.equals("bye")) {
+                    ui.printBye();
+                    isExit = true;
+                } else if (instruction.equals("list")) {
+                    taskList.printList();
+                } else if (instruction.equals("done")) {
+                    int index = Parser.parseIndex(input);
+                    taskList.markTask(index);
+                    storage.writeSavedList(taskList.getList());
+                } else if (instruction.equals("delete")) {
+                    int index = Parser.parseIndex(input);
+                    taskList.deleteTask(index);
+                    storage.writeSavedList(taskList.getList());
+                } else if (instruction.equals("find")) {
+                    String searchPhrase = Parser.parseSearchPhrase(input);
+                    ui.printFoundTasks(taskList.findTasks(searchPhrase));
+                } else if (instruction.equals("todo") || instruction.equals("deadline")
+                        || instruction.equals("event")) {
+                    try {
+                        if (instruction.equals("todo")) {
+                            String taskDescription = Parser.parseDescription(input, true);
+                            ui.testEmptyDescription(taskDescription);
+                            taskList.createToDo(taskDescription);
+                            storage.writeSavedList(taskList.getList());
+                        } else {
+                            String taskDescription = Parser.parseDescription(input, false);
+                            ui.testTimeFormat(taskDescription);
+                            String taskContent = Parser.parseContent(taskDescription);
+                            LocalDateTime taskTime = Parser.parseTime(taskDescription);
+                            ui.testEmptyDescription(taskContent);
```

Comment:
`testEmptyDescription` doesn't really tell me what's going on, and doesn't seem like it throws an error. I had to go to the method declaration to see what it does.

maybe 

```java
if (somethingElse.isEmptyDescription(taskContent)) {
  throw new Exception(...);
}
```
would be better here as the reader can see what's going on.

If you would like to keep the exception inside the method i guess
```
throwIfEmptyDescription(...);
```
would be more descriptive as well;

Also don't think it should be part of ui.